### PR TITLE
Constants as first-class citizens

### PR DESCRIPTION
**Pull request by ditansu**
Making the constants first-class citizens. Support Builtin, User, and Package.

was
``` fsharp
let fns : List<BuiltInFn> =
  [ { name = fn "Math" "pi" 0
      typeParams = []
      parameters = []
      returnType = TFloat
      description =
        "Returns an approximation for the mathematical constant {{π}}, the ratio of a
         circle's circumference to its diameter."
      fn =
        (function
        | _, _, [] -> Ply(DFloat System.Math.PI)
        | _ -> incorrectArgs ())
      sqlSpec = NotYetImplemented
      previewable = Pure
      deprecated = NotDeprecated }
...
```
became
```fsharp

let constants : List<BuiltInConstant> =
  [ { name = constant "Math" "pi" 0
      returnType = TFloat
      description =
        "Returns an approximation for the mathematical constant {{π}}, the ratio of a
         circle's circumference to its diameter."
      constant = Ply(DFloat System.Math.PI)
      sqlSpec = NotYetImplemented
      previewable = Pure
      deprecated = NotDeprecated }
...
```

### CODE REVIEW

File: backend/src/LibBackend/SqlCompiler.fs [hunk 1]
```diff
@@ -260,10 +261,24 @@ let rec lambdaToSql
...
+      | Some c ->
+        typecheck nameStr c.returnType expectedType
+        let random = randomString 8
+        let newname = $"{nameStr}_{random}"
+        // let! constant = c.constant
+        let (sqlValue, actualType) = dvalToSql expectedType c.constant.Result
+        $"(@{newname})", [ newname, sqlValue ], actualType
+      | None ->
+        error
+          $"Only builtin constants can be used in queries right now; {nameStr} is not a builtin constant"
```
Comment (pbiggar) [inline backend/src/LibBackend/SqlCompiler.fs:280]: It seems like we should be able to use non-builtin constants, but if that's hard to swing then that can be something to follow up with.

File: backend/src/StdLibCliInternal/Libs/Cli.fs [hunk 2]
```diff
@@ -17,18 +17,21 @@ let typ = FQTypeName.stdlibTypeName'
...
 
-let (stdlibFns, stdlibTypes) =
+let (stdlibFns, stdlibTypes, stdlibConstants) =
   LibExecution.StdLib.combine
     [ StdLibExecution.StdLib.contents; StdLibCli.StdLib.contents ]
     []
     []
 
-
+// TODO: think about add constants
```
Comment (pbiggar) [inline backend/src/StdLibCliInternal/Libs/Cli.fs:26]: Unclear what this comment is saying.

File: backend/src/LibExecution/RuntimeTypes.fs [hunk 3]
```diff
@@ -884,6 +966,15 @@ type BuiltInType =
...
     deprecated : Deprecation<FQTypeName.StdlibTypeName> }
 
+type BuiltInConstant =
+  { name : FQConstantName.StdlibConstantName
+    returnType : TypeReference
+    description : string
+    previewable : Previewable
+    deprecated : Deprecation<FQConstantName.StdlibConstantName>
+    sqlSpec : SqlSpec
+    constant : DvalTask }
```
Comment (pbiggar) [inline backend/src/LibExecution/RuntimeTypes.fs:976]: I don't think this needs to be a DvalTask. I think a Dval should be fine

File: backend/src/LibExecution/RuntimeTypes.fs [hunk 4]
```diff
@@ -884,6 +966,15 @@ type BuiltInType =
     description : string
     deprecated : Deprecation<FQTypeName.StdlibTypeName> }
 
+type BuiltInConstant =
+  { name : FQConstantName.StdlibConstantName
+    returnType : TypeReference
+    description : string
+    previewable : Previewable
```
Comment (pbiggar) [inline backend/src/LibExecution/RuntimeTypes.fs:973]: Constants should all be Previewable so this field is probably not necessary

File: backend/src/LibExecution/RuntimeTypes.fs [hunk 5]
```diff
@@ -884,6 +966,15 @@ type BuiltInType =
     description : string
     deprecated : Deprecation<FQTypeName.StdlibTypeName> }
 
+type BuiltInConstant =
+  { name : FQConstantName.StdlibConstantName
+    returnType : TypeReference
```
Comment (pbiggar) [inline backend/src/LibExecution/RuntimeTypes.fs:971]: Returntype should probably be called `typ`

File: backend/src/LibExecution/RuntimeTypes.fs [hunk 6]
```diff
@@ -884,6 +966,15 @@ type BuiltInType =
...
+    description : string
+    previewable : Previewable
+    deprecated : Deprecation<FQConstantName.StdlibConstantName>
+    sqlSpec : SqlSpec
```
Comment (pbiggar) [inline backend/src/LibExecution/RuntimeTypes.fs:975]: I don't think we need an sqlSpec for this, as the value is known.

File: backend/src/LibExecution/ProgramTypesParser.fs [hunk 7]
```diff
@@ -80,4 +80,5 @@ module Toplevel =
     | PT.Toplevel.TLDB _ -> "db"
     | PT.Toplevel.TLHandler _ -> "handler"
     | PT.Toplevel.TLFunction _ -> "user_function"
-    | PT.Toplevel.TLType _ -> "user_tipe"
+    | PT.Toplevel.TLType _ -> "user_tipe" // t`i`pe typo?
```
Comment (pbiggar) [inline backend/src/LibExecution/ProgramTypesParser.fs:83]: This isn't a typo, but good point - we can change that to `user_type` now

**Review by pbiggar — CHANGES_REQUESTED** [review]
This looks great, I'd say 99% of the way there!

**Comment by ditansu** [comment]
The tons of conflicts were resolved, spent a lot of time 🫠

**Comment by ditansu** [comment]
@StachuDotNet Oh, no problem, thanks. I can do it myself :) yes, it's hard but, this way I keep an eye on changes.

**Comment by StachuDotNet** [comment]
Please let me know if it gets to be too much to rebase - a lot of has been moving and I feel kinda bad for you having to deal with it :sweat_smile:

**Comment by pbiggar** [comment]
fyi, I'm about to make a bunch of big changes, so I'm going to try to merge this PR this morning.

**Comment by pbiggar** [comment]
Thanks again for the massive effort on this, @ditansu! It's a massive PR and the constant changes couldn't have made this fun. We appreciate all the work it took!

**Comment by ditansu** [comment]
Thanks for warm words. It's took a more time than I guessed. I hope next time I solve an issue faster.

File: backend/src/LibExecution/ProgramTypes.fs [hunk 8]
```diff
@@ -505,6 +576,26 @@ module UserType =
...
 
+
+
+type Const =
+  | CInt of int64
+  | CBool of bool
+  | CString of string
+  | CChar of string
+  | CFloat of Sign * string * string
+  | CUnit
```
Comment (StachuDotNet) [inline backend/src/LibExecution/ProgramTypes.fs:587]: I wonder - when would I use a Const of a Unit?

Comment (pbiggar) [reply backend/src/LibExecution/ProgramTypes.fs:587]: Perhaps

```
let myError = Error ....
let mySuccess = Ok ()
```

Comment (StachuDotNet) [reply backend/src/LibExecution/ProgramTypes.fs:587]: Good point - thanks!

**Pull request merged**